### PR TITLE
feat(electric): Use two replications slots to allow clients to resume replication after Electric restarts

### PR DESCRIPTION
**Pull request by alco**
Continuing on from https://github.com/electric-sql/electric/pull/1099, this PR introduces the concept of a "persistent replication slot". Now, instead of having an auto-advancing replication slot, Electric can keep around WAL records that contain transactions it has already seen and acknowledged previously. As a result, when Electric restarts, it can repopulate its in-memory cache of transactions and be able to resume clients' replication streams.

New configuration options `ELECTRIC_TXN_CACHE_SIZE` and `ELECTRIC_RESUMABLE_WAL_WINDOW` have been added. The relevant docs are updated in https://github.com/electric-sql/electric/pull/1050.

### CODE REVIEW

File: components/electric/lib/electric/config.ex [hunk 1]
```diff
@@ -313,4 +319,62 @@ defmodule Electric.Config do
   defp parse_database(nil, username), do: username
   defp parse_database("/", username), do: username
   defp parse_database("/" <> dbname, _username), do: dbname
+
+  @spec parse_human_readable_size(binary, pos_integer) :: {:ok, pos_integer} | {:error, binary}
```
Comment (icehaunter) [inline components/electric/lib/electric/config.ex:323]:
```suggestion
  @spec parse_human_readable_size(binary | nil, pos_integer) :: {:ok, pos_integer} | {:error, binary}
```

Slightly wrong spec

File: components/electric/lib/electric/replication/postgres/logical_replication_producer.ex [hunk 2]
```diff
@@ -324,9 +360,30 @@ defmodule Electric.Replication.Postgres.LogicalReplicationProducer do
...
     Logger.debug("Acknowledging #{lsn}", origin: origin)
     Client.acknowledge_lsn(conn, lsn)
   end
+
+  # Advance the replication slot to let Postgres discard old WAL records.
+  #
+  # TODO: make sure we're not removing transactions that are about to be requested by a newly
+  # connected client. See VAX-1552.
+  #
+  # TODO(optimization): do not run this after every consumed transaction.
```
Comment (icehaunter) [inline components/electric/lib/electric/replication/postgres/logical_replication_producer.ex:374]: So I have 2 questions here:

1. Why are we running this whenever we see a `Commit`? We have `ack_fn` set to `ack` being the previous function - I believe it makes more sense to advance the slot as we do right now - once stored in memory at least (although I'll agree that it matters a bit less given that we're advancing a secondary slot here).
2. I see the "optimization" note you've left here, but I'm not sure I'm comfortable leaving this as a TODO. Advancing a slot on every transaction is fine-ish, however we're opening a new fully separate connection to PG on every received transaction. This is going to be extremely slow and costly for us, and PG won't be able to keep up with the amount of connections we're opening for one-shot queries. I'd be against even a pooling approach here - I believe we need to just open and hold a stable connection to PG to perform these tasks.

File: components/electric/lib/electric/replication/postgres/slot_server.ex [hunk 3]
```diff
@@ -371,7 +373,7 @@ defmodule Electric.Replication.Postgres.SlotServer do
          %Changes.Transaction{commit_timestamp: ts, changes: changes, origin: origin},
          %State{} = state
        ) do
-    first_lsn = Lsn.increment(state.current_lsn)
+    first_lsn = Lsn.increment(state.current_lsn, @lsn_step)
```
Comment (icehaunter) [inline components/electric/lib/electric/replication/postgres/slot_server.ex:376]: What's the reason to introduce these steps?

**Review by icehaunter — CHANGES_REQUESTED** [review]
Great work, exactly what I had in mind myself, and additional thanks for cleaning and testing the LSN module. One big question from me

File: components/electric/lib/electric/replication/postgres/logical_replication_producer.ex (hunk 2 above)
Comment (alco) [reply components/electric/lib/electric/replication/postgres/logical_replication_producer.ex:374]: The optimization note wasn't meant to be left in the code for long. I split the code into multiple PRs to help myself draw boundaries between changes to different components and to get to a working version faster to then be able to focus on the edge cases and inefficiencies.

So, point taken. Holding a persistent connection might be the way to go if the current approach of advancing the slot often remains in place. Though I'm hoping to implement more of a batched approach, e.g. advancing every N transactions or in size-on-disk increments, at which point relying on a connection pool may be sufficient. I'll circle back to this discussion when I have more concrete ideas to share but I'm expecting those to be implemented in a follow-up PR in any case.

---

N.B.: I had a connection pool implemented as part of the early implementation work but had to take that out for now just to keep the scope of changes that are already spread over multiple PRs manageable.

Comment (icehaunter) [reply components/electric/lib/electric/replication/postgres/logical_replication_producer.ex:374]: With or without batching, holding a single open connection just for slot advancement is fine, since only one process is likely to ever do logical replication stream consumption. I think just spawn a sibling GenServer that holds the advancing connection - to split the responsibility a little - and just `cast` to that server on `ack` - this makes the whole approach easy and doesn't clog up logical replication consumer with ack message handling since `cast`s will be sent to a different process. And if that genserver goes down and we lose a couple casts in the middle - no problem, worst case we'll have slightly more WAL stored at a point in time, but it'll resolve on any next cast. 

I'm not sure how you want to structure the PR merge sequence, but I would insist on adding this functionality in this PR. It's not much code, but the difference in perfomance implications is huge.

Comment (alco) [reply components/electric/lib/electric/replication/postgres/logical_replication_producer.ex]: The implementation has changed now that this PR is based on https://github.com/electric-sql/electric/pull/1099 which is a fix for https://github.com/electric-sql/electric/issues/1083.

See the new commit - https://github.com/electric-sql/electric/pull/1043/commits/76addc4da4b5a4f5d092dfaace65fc964b99e3d5.

TL;DR: LogicalReplicationProducer now opens a persistent "service connection" during initialization and uses it to advance the main slot with the help of a timer.

File: components/electric/lib/electric/postgres/lsn.ex [hunk 4]
```diff
@@ -1,62 +1,194 @@
...
   end
 
   defimpl String.Chars do
-    alias Electric.Postgres.Lsn
-    def to_string(lsn), do: Lsn.to_string(lsn)
+    def to_string(lsn), do: "#{Electric.Postgres.Lsn.to_iolist(lsn)}"
+  end
+
+  defimpl List.Chars do
+    def to_charlist(lsn), do: ~c'#{Electric.Postgres.Lsn.to_iolist(lsn)}'
```
Comment (magnetised) [inline components/electric/lib/electric/postgres/lsn.ex:192]: nitpick: do you need to wrap this with the `~c` sigil?

Comment (alco) [reply components/electric/lib/electric/postgres/lsn.ex:192]: Mix formatter has been adding those since [Elixir v1.15.0](https://github.com/elixir-lang/elixir/blob/v1.15/CHANGELOG.md#v1150-2023-06-19):
> [Code] Code.format_string!/2 now converts 'charlists' into ~c"charlists" by default

**Review by magnetised — APPROVED** [review]
Super

File: components/electric/lib/electric/replication/postgres/logical_replication_producer.ex [hunk 5]
```diff
@@ -334,21 +411,45 @@ defmodule Electric.Replication.Postgres.LogicalReplicationProducer do
...
+    repl_conn = state.repl_conn
     origin = state.origin
 
     %Transaction{
       transaction
       | lsn: end_lsn,
         # Make sure not to pass state.field into ack function, as this
         # will create a copy of the whole state in memory when sending a message
-        ack_fn: fn -> ack(conn, origin, end_lsn) end
+        ack_fn: fn -> ack(repl_conn, origin, end_lsn) end
```
Comment (icehaunter) [inline components/electric/lib/electric/replication/postgres/logical_replication_producer.ex:422]: I think this `ack` should not operate on `repl_conn`. 

`ack_fn` function is going to be called in a different process later down the line - in particular, in WAL cache consumer stage. This means that if this repl connection is closed for any reason, old transcations will call it with a stale value and probably crash - I don't know how `:epgsql.standby_status_update/3` acts when it's passed a nonexistent pid. 

To decouple that a little and make sure that the process calling the `ack_fn` doesn't crash, I'd rather `ack_fn` sent a `GenServer.cast/2` to a process that would actually hold the repl conn and do the acknowledge. If the cast fails because the holding GenServer is temporarily dead -- no problem, main code continues to operate, and worst case we acknowledge on the next magic write anyway. So I'd rather put this "advancing" repl conn into a separate GenServer that can restart and reconnect separately, and just `GenServer.cast` into it - maybe make it a `:gproc` via-tuple using the origin so that we don't even care where the repl process currently lives. 

What do you think?

**Review by icehaunter — CHANGES_REQUESTED** [review]

File: components/electric/lib/electric/postgres/cached_wal/ets_backed.ex [hunk 6]
```diff
@@ -182,19 +182,30 @@ defmodule Electric.Postgres.CachedWal.EtsBacked do
...
+    # https://www.postgresql.org/message-id/E1nZNz3-001zFN-UA%40gemulon.postgresql.org
     |> Stream.each(
       &Logger.debug(
         "Saving transaction #{&1.xid} at #{&1.lsn} with changes #{inspect(&1.changes)}"
       )
     )
-    |> Stream.map(fn %Transaction{lsn: lsn} = tx ->
-      {lsn_to_position(lsn), %{tx | ack_fn: nil}}
-    end)
+    |> Stream.map(fn %Transaction{} = tx -> {lsn_to_position(tx.lsn), tx} end)
```
Comment (icehaunter) [inline components/electric/lib/electric/postgres/cached_wal/ets_backed.ex:208]: nitpick: I don't understand why you don't like destructuring assignment of `lsn`, given that it's faster than dynamic dispatch of a `.` operator. Doesn't really matter in this case, just caught my eye.

File: components/electric/lib/electric/replication/postgres/logical_replication_producer.ex [hunk 7]
```diff
@@ -339,37 +374,28 @@ defmodule Electric.Replication.Postgres.LogicalReplicationProducer do
...
-  defp emit_events(state, []) do
-    {:noreply, [], state}
-  end
+  defp dispatch_events(%{demand: demand, queue: queue, queue_len: queue_len} = state) do
+    {to_emit, queue_remaining} = :queue.split(demand, queue)
+    to_emit |> :queue.last() |> ack(state)
 
-  defp emit_events(state, events) do
-    {:noreply, Enum.reverse(events), state}
+    state = %{state | queue: queue_remaining, queue_Len: queue_len - demand, demand: 0}
```
Comment (icehaunter) [inline components/electric/lib/electric/replication/postgres/logical_replication_producer.ex:397]: Typo that should lead to a runtime error, possibly means that this code path is not covered by test (although this is caught by Dialyzer)
```suggestion
    state = %{state | queue: queue_remaining, queue_len: queue_len - demand, demand: 0}
```

**Review by icehaunter — APPROVED** [review]
Nice design, I like it! a couple of nitpicks, but nothing more. Let's coordinate a release before merging this though

File: components/electric/lib/electric/postgres/cached_wal/ets_backed.ex (hunk 6 above)
Comment (alco) [reply components/electric/lib/electric/postgres/cached_wal/ets_backed.ex:208]: It's not that I don't like it. As a rule of thumb, I use pattern matching in function heads when the same field is used multiple times in the function body. Because we have instances in the code where some fields are pattern-matched while others are accessed via the dot operator, the split usually looks arbitrary.

I believe that BEAM is smart enough to optimize uses of the dot operator if there's pattern-matching/guard in the function head. Can't find the source now, but here's at least a partial confirmation of that https://elixirforum.com/t/way-too-much-detail-about-matching-in-the-head-vs-accessing-maps-in-the-body/49167/2

File: components/electric/lib/electric/replication/postgres/logical_replication_producer.ex (hunk 7 above)
Comment (alco) [reply components/electric/lib/electric/replication/postgres/logical_replication_producer.ex:397]: Yeah, I had to stop working on this abruptly yesterday and pushed what I had ready by then. This condition requires a specific test setup to cover. I think having dialyzer complain about it is good enough.

File: components/electric/lib/electric/replication/postgres/logical_replication_producer.ex (hunk 5 above)
Comment (alco) [reply components/electric/lib/electric/replication/postgres/logical_replication_producer.ex]: As we've discussed elsewhere, it is fine to acknowledge transactions inside LogicalReplicationProducer and rely on Elixir's supervision tree to restart things if anything fails. Since we now have a replication slot with a persistent starting point, we can replay already-seen transactions when a new replication connection opens.
